great_expectations/profiler/profiler_rule/rule_state.py

```diff
@@ -0,0 +1,134 @@
+DOMAIN_KWARGS_PARAMETER_FULLY_QUALIFIED_NAME: str = (
+    f"$domain.{DOMAIN_KWARGS_PARAMETER_NAME}"
+)
+VARIABLES_KEY: str = "$variables."
+
+
+# TODO: <Alex>ALEX -- We should consider changing the names "parameters" and "Parameter.parameters" to be more semantically meaningful.</Alex>
+class RuleState:
+    """Manages state for ProfilerRule objects. Keeps track of rule domain, rule parameters,
+    and any other necessary variables for validating the rule."""
+
+    # TODO: <Alex>ALEX -- Add type hints; what are the types?</Alex>
+    def __init__(
+        self,
+        active_domain: Optional[
+            Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]
+        ] = None,
+        domains: Optional[
+            List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]]
+        ] = None,
+        # TODO: <Alex>ALEX -- what is the structure of this "parameters" argument?</Alex>
+        parameters: Optional[Dict[str, Parameter]] = None,
+        variables: Optional[Parameter] = None,
+    ):
+        self._active_domain = active_domain
+        if domains is None:
+            domains = {}
+        self._domains = domains
+        if parameters is None:
+            parameters = {}
+        self._parameters = parameters
+        # TODO: <Alex>ALEX -- what is the type -- what kind of a dictionary is "variables"?  Is is "Parameter"?</Alex>
+        if variables is None:
+            variables = Parameter(parameters={}, details=None)
+        self._variables = variables
+
+    @property
+    def parameters(self) -> Dict[str, Parameter]:
+        return self._parameters
+
+    @property
+    def domains(self) -> List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]]:
+        return self._domains
+
+    @domains.setter
+    def domains(
+        self, domains: List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]]
+    ):
+        self._domains = domains
+
+    @property
+    def active_domain(self) -> Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]:
+        return self._active_domain
+
+    @active_domain.setter
+    def active_domain(
+        self, active_domain: Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]
+    ):
+        self._active_domain = active_domain
+
+    # TODO: <Alex>ALEX -- what is the return type?</Alex>
+    @property
+    def variables(self) -> Parameter:
+        """
+        Getter for rule_state variables
+        :return: variables necessary for validating rule
+        """
+        return self._variables
+
+    @property
+    def active_domain_id(self) -> str:
+        """
+        Getter for the id of the rule domain
+        :return: the id of the rule domain
+        """
+        return IDDict(self.active_domain).to_id()
+
+    def get_parameter_value(self, parameter_name: str) -> Any:
+        """
+        Get a value from the current rule state. Values must be dot-delimited, and may start either with
+        the key 'domain' or the id of a parameter.
```

Comment:
Is the id of a parameter meaningful? I thought that was the unique identifier that we had looked at? Or does this mean more like `$mean`? Are `$mean` and that long unique identifier both parameter IDs?

Reply:
@talagluck This is a great question.  We see that `ParameterBuilder` takes `parameter_id` as a parameter in its constructor.  Then we see that `ProfilerRule.evaluate()` uses it as follows: `rule_state.parameters[domain_id][parameter_id] = parameter.parameters`.  However, I have not yet seen `ParameterBuilder` (or any of its descendants) being instantiated.  According to `https://great-expectations--2696.org.readthedocs.build/en/2696/guides/how_to_guides/creating_and_editing_expectations/how_to_create_an_expectation_suite_with_the_user_configurable_profiler.html`, all parameter builder configurations have `id` and `class_name` in common, but then use specific attributes pertaining the logic of each particular parameter builder.  Hence, I will change the constructor argument to `id`, but otherwise will wait for further discussion to agree on the next steps.  Thank you.

Reply:
I made a change that proposes that `parameter_id` is not meaningful and that `parameter_builder` should use `parameter_name` instead of `id`.